component/azstorage/block_blob.go

```diff
@@ -766,16 +772,98 @@ func (bb *BlockBlob) createNewBlocks(blockList *common.BlockOffsetList, offset,
 	// BufferSize is the size of the buffer that will go beyond our current blob (appended)
 	var bufferSize int64
 	for i := prevIndex; i < offset+length; i += blockSize {
-		// create a new block if we hit our block size
 		blkSize := int64(math.Min(float64(blockSize), float64((offset+length)-i)))
-		newBlock := bb.createBlock(blockIdLength, i, blkSize)
+		newBlock := bb.createBlock(blockList.BlockIdLength, i, blkSize)
 		blockList.BlockList = append(blockList.BlockList, newBlock)
-		// reset the counter since it will help us to determine if there is leftovers at the end
+		// reset the counter to determine if there are leftovers at the end
 		bufferSize += blkSize
 	}
 	return bufferSize
 }
 
+func (bb *BlockBlob) removeBlocks(blockList *common.BlockOffsetList, size int64, name string) *common.BlockOffsetList {
+	_, index := blockList.BinarySearch(size)
+	// if the start index is equal to new size - block should be removed - move one index back
+	if blockList.BlockList[index].StartIndex == size {
+		index = index - 1
+	}
+	// if the file we're shrinking is in the middle of a block then shrink that block
+	if blockList.BlockList[index].EndIndex > size {
+		blk := blockList.BlockList[index]
+		blk.EndIndex = size
+		blk.Data = make([]byte, blk.EndIndex-blk.StartIndex)
+		blk.Flags.Set(common.DirtyBlock)
+		bb.ReadInBuffer(name, blk.StartIndex, blk.EndIndex-blk.StartIndex, blk.Data)
+	}
+
+	blockList.BlockList = blockList.BlockList[:index+1]
+	return blockList
+}
+
+func (bb *BlockBlob) TruncateFile(name string, size int64) error {
+	// log.Trace("BlockBlob::TruncateFile : name=%s, size=%d", name, size)
+	attr, err := bb.GetAttr(name)
+	if err != nil {
+		log.Err("BlockBlob::TruncateFile : Failed to get attributes of file %s (%s)", name, err.Error())
+		if err == syscall.ENOENT {
+			return err
+		}
+	}
+	if size == 0 || attr.Size == 0 {
+		err := bb.WriteFromBuffer(name, nil, make([]byte, size))
```

Comment:
Do we need to worry about size being huge and allocating a huge array?

Reply:
yes, that is a possibility - it's one of the improvement items post this PR. Did not want to have too many changes in this PR.